torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,159 @@ namespace rpc {
+
+    steady_clock_time_point earliestTimeout;
 
-RpcAgent::~RpcAgent() = default;
+    if (!rpcRetryMap_.empty()) {
+      earliestTimeout = rpcRetryMap_.begin()->first;
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
+        return !(rpcAgentRunning_.load());
+      });
+    }
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+
+    // Updating these since something may have been added to the map while this
+    // thread was sleeping.
+    earliestTimeout = rpcRetryMap_.begin()->first;
+    auto& earliestRpcList = rpcRetryMap_.begin()->second;
+
+    // We iterate through all the RPC's set to be retried at the current
+    // timepoint, resend those RPC's, and add a callback that either schedules
+    // the RPC for a future retry or marks it with success/error depending on
+    // the outcome of the current send. Then, we clean up the rpcRetryMap_.
+    for (auto it = earliestRpcList.begin(); it != earliestRpcList.end();
+         /* no increment */) {
+      auto& earliestRpc = *it;
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
+      earliestRpc->retryCount_++;
+
+      fm->addCallback([this, newTime, earliestRpc](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        rpcRetryCallback(message, futErr, newTime, earliestRpc);
+      });
+
+      // We have either scheduled the RPC for a future retry, or we've marked
+      // it complete upon success, or we've marked it with an error after
+      // maxRetries unsuccessful send attempts. Thus, we can remove it from the
+      // retry Map for the current timepoint.
+      it = earliestRpcList.erase(it);
+    }
+
+    // If there are no more RPC's set to be retried at the current timepoint,
+    // we can remove the corresponsing unordered_set from the retry map.
+    if (earliestRpcList.empty()) {
+      rpcRetryMap_.erase(earliestTimeout);
+    }
+  }
+}
+
+void RpcAgent::rpcRetryCallback(
+    const rpc::Message& message,
+    const c10::optional<utils::FutureError>& futErr,
+    steady_clock_time_point newTime,
+    std::shared_ptr<RpcRetryInfo> earliestRpc) {
+  if (futErr) {
+    // Adding one since we want to include the original send as well and not
+    // just the retry count.
+    LOG(INFO) << "Send try " << std::to_string(earliestRpc->retryCount_ + 1)
+              << " failed";
+    if (earliestRpc->retryCount_ < earliestRpc->maxRetries_) {
+      // If the previous future completed with an error and we haven't
+      // completed maxRetries send attempts, we move the earliestRpc
+      // struct to a new time point in the retry map (effectively
+      // scheduling it for a future retry.)
+      {
+        std::lock_guard<std::mutex> retryMapLock(rpcRetryMutex_);
+        rpcRetryMap_[newTime].emplace(std::move(earliestRpc));
```

Comment:
I am trying to understand the logic here. The retry is only inserted when the previous future is either timeout or failed, so the retry delay is actually `rpc timeout + backoff`, right? And this is also why we don't need to worry about marking a future as complete multiple times?

Reply:
Right, since we moved all scheduling of RPC's to the callbacks, there is only one outstanding send for a given RPC at any given time, so there's no need to check whether the original future has already been marked complete. This change also causes the effective retry duration to be `rpc timeout + backoff` as you mentioned, since we wait for the previous send to fail before the callback schedules the next try.